src/capture/image/image_decoder.h

```diff
@@ -106,32 +106,24 @@ class ImageDecoder {
 
 		switch (image.params.pixel_format) {
 		case PixelFormat::RGB555_Packed16: {
-			const auto p = read_unaligned_uint16(pos);
+			const auto p = host_readw(pos);
```

Comment:
the `host_readx` functions pass through the functions to adjust for the host endianness :

```cpp
// Read a 16-bit word from 8-bit DOS/little-endian byte-ordered memory.
static inline uint16_t host_readw(const uint8_t* const arr) noexcept
{
	return le16_to_host(read_unaligned_uint16(arr));
}
```

Reply:
Thanks @kklobe for the refresher.

I just checked git blame because I thought this code looked familiar.  I made this change a couple months ago citing an endinaess bug.  97755b29d2bc5ef81ccfd544c372431f2eae2629

My change is essentially being reverted here.  Has the image capturer been tested on big endian to confirm this is correct?

Reply:
Yup, I tested image capture (raw scaled and post-rendered) at all colour depths and video modes, including this one (using sdd553a), and it was broken on big-endian.

It was working backward from the image capture colour values (I believe from raw scaled code path) that eventually landed me here :-)

When I read your documentation about this the RGB555_Packed16 enum:

> "stored as packed uint16 data with highest bit unused in host native endianess."

This change was like a puzzle piece, given the change was flipping the DOS-written little-endian 16-bit value (from the emulated S3 VESA memory space) over to a host-native value, which is exactly as you've described.

Reply:
> My change is essentially being reverted here. Has the image capturer been tested on big endian to confirm this is correct?

Just confirming, your commit message says "Fix endianness bug in image decoder". Do you remember, did you fix an actual observed problem, or you thought you're fixing an endianness bug, but were not able to actually test it on BE?

If the latter, that explains the situation. Just wanna get to the bottom of it to 100% understand the situation, that's all 😄

Reply:
Btw, I find this function confusing as fuck, as _most_ BE/LE conversion functions in most libraries. I would need to draw to understand it, like always, but if this is what I think it is based on the comment ("Read a 16-bit word from 8-bit DOS/little-endian byte-ordered memory."), `read_from_16le` would be a far better name IMO. At least that's how my brain works 😄 

```ini
static inline uint16_t read_from_16le(const uint8_t* const arr) noexcept
``

Reply:
@weirddan455 All good man, the important thing is we get it right finally right 😅 This is a very confusing topic, indeed, and you've cleaned up & clarified the rest of the pixel formats, so your success rate is pretty high even with getting a single one wrong 😄

Reply:
> Then these read functions get byte pointer args. Similarly for writes, `write_as_16le` starting at the byte-pointer, etc.

> Also, agree that something like `read_from_16le` would be a better name.

@johnnovak , @weirddan455 -> :+1: 

I added the _unaligned_ accessors early in 2020 - 505669046719430186a057ae3f7d5d0033c6a689:

- `read_unaligned_*`
- `read_unaligned_*_at`
- `write_unaligned_*`
- `write_unaligned_*_at`
- `inc_unaligned_*`
- `add_to_unaligned_*`

But these don't involve big/little endian swaps. The `host_read*` and `host_write*` calls have been in the code since 2003.

A bit of context for the `host_`, from my understanding:

DOSBox juggles two worlds, the native host platform space where the business logic and emulation happens, and the emulated x86 hardware and DOS-space, which is always LE.

If we need to pluck values from DOS space into host-space and actually work with them (ie: compare them, check if they're beyond 32000) then it's done with these `host_read` calls.  Think of them as _"the host is reading a value from DOS-land"_.

The `host_write` takes a value from the host and writes it back into DOS memory. Think of it like, _"The host is writing a value into DOS-land"_.

It's likely that the names were chosen through the eyes of DOS, so "the host" is the thing acting on it. 

If we flipped that around to be "through the eyes of the host", we could be the only acting on the DOS space, with calls like: `read_uint32_from_dos` and `write_uint16_to_dos`. 

Both angles-of-attack make the code a bit more self-documenting, and there's probably room to mass-rename them or simply add more aliases, because it might make it clearer.

The big endian fixups are a nice way to catch these boundaries between the host world and DOS world, where previously the code might be more ambiguously working with memory and values - and it becomes hard to tell which "world" we're in.

Reply:
I kinda get the reasoning behind it, but I'd prefer if these were explicit and that's it. So `read_from_16le` and `write_as_16le`, etc. and none of this "host" concept. Then of course anyone messing around with a DOS emulator must 100% know that DOS PCs were a strictly little-endian platform, and job done 😎

Reply:
Our of scope for this PR, but it's only 200+ touch points in the code for a mass-rename PR. 

But anyone porting features to/from Staging and other DOSBox's will have to be aware of this (most of the big community patches are in, so this risk is mostly moot anyway). 

```
[dosbox-staging] (main $) git grep -c 'host_read\|host_write'
include/mem.h:11
include/mem_host.h:30
include/paging.h:12
src/capture/capture_audio.cpp:4
src/capture/capture_video.cpp:7
src/cpu/core_dyn_x86/decoder.h:2
src/cpu/core_dynrec/decoder_basic.h:6
src/cpu/core_simple.cpp:3
src/cpu/dyn_cache.h:12
src/cpu/paging.cpp:6
src/dos/dos_execute.cpp:2
src/dos/dos_keyboard_layout.cpp:16
src/dos/dos_tables.cpp:1
src/dos/program_boot.cpp:2
src/hardware/gus.cpp:1
src/hardware/ide.cpp:61
src/hardware/mixer.cpp:12
src/hardware/vga_memory.cpp:24
```